moveit_ros/perception/handeye_calibration_target/src/handeye_aruco_target.cpp

```diff
@@ -0,0 +1,282 @@
+                                                  int border_bits, const std::string& dictionary_id)
+{
+  if (markers_x <= 0 || markers_y <= 0 || marker_size <= 0 || separation <= 0 || border_bits <= 0 ||
+      marker_dictionaries_.find(dictionary_id) == marker_dictionaries_.end())
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Invalid target intrinsic params.\n"
+                                        << "markers_x_ " << std::to_string(markers_x) << "\n"
+                                        << "markers_y_ " << std::to_string(markers_y) << "\n"
+                                        << "marker_size " << std::to_string(marker_size) << "\n"
+                                        << "separation " << std::to_string(separation) << "\n"
+                                        << "border_bits " << std::to_string(border_bits) << "\n"
+                                        << "dictionary_id " << dictionary_id << "\n");
+    return false;
+  }
+
+  std::lock_guard<std::mutex> aruco_lock(aruco_mutex_);
+  markers_x_ = markers_x;
+  markers_y_ = markers_y;
+  marker_size_ = marker_size;
+  separation_ = separation;
+  border_bits_ = border_bits;
+
+  const auto& it = marker_dictionaries_.find(dictionary_id);
+  dictionary_id_ = it->second;
+
+  return true;
+}
+
+bool HandEyeArucoTarget::setTargetDimension(double marker_measured_size, double marker_measured_separation)
+{
+  if (marker_measured_size <= 0 || marker_measured_separation <= 0)
+  {
+    ROS_ERROR_NAMED(LOGNAME, "Invalid target measured dimensions: marker_size %f, marker_seperation %f",
+                    marker_measured_size, marker_measured_separation);
+    return false;
+  }
+
+  std::lock_guard<std::mutex> aruco_lock(aruco_mutex_);
+  marker_size_real_ = marker_measured_size;
+  marker_separation_real_ = marker_measured_separation;
+  ROS_INFO_STREAM_NAMED(LOGNAME, "Set target real dimensions: \n"
+                                     << "marker_measured_size " << std::to_string(marker_measured_size) << "\n"
+                                     << "marker_measured_separation " << std::to_string(marker_measured_separation)
+                                     << "\n");
+  return true;
+}
+
+bool HandEyeArucoTarget::createTargetImage(cv::Mat& image) const
+{
+  cv::Size image_size;
+  image_size.width = markers_x_ * (marker_size_ + separation_) - separation_ + 2 * separation_;
+  image_size.height = markers_y_ * (marker_size_ + separation_) - separation_ + 2 * separation_;
+
+  try
+  {
+    // Create target
+    cv::Ptr<cv::aruco::Dictionary> dictionary = cv::aruco::getPredefinedDictionary(dictionary_id_);
+    cv::Ptr<cv::aruco::GridBoard> board =
+        cv::aruco::GridBoard::create(markers_x_, markers_y_, float(marker_size_), float(separation_), dictionary);
+
+    // Create target image
+    board->draw(image_size, image, separation_, border_bits_);
+  }
+  catch (cv::Exception& e)
```

Comment:
Catch by const ref. Not because it changes anything in the compiler generated code but because it signals to other developers that you doing what they would expect you to do, which is just reading the exception value instead of modifying it.

Reply:
Good point! `const` situations should be paid more attention, otherwise some undefined behavior could be spread elsewhere. Changed this to const ref.